cli/src/main/java/org/aya/cli/literate/AyaMdParser.java

```diff
@@ -0,0 +1,232 @@
+        yield children.sizeEquals(1) ? children.first() : new Literate.Many(null, children, true);
+      }
+      case FencedCodeBlock codeBlock -> {
+        var language = codeBlock.getInfo();
+        var raw = codeBlock.getLiteral();
+        var spans = codeBlock.getSourceSpans();
+        if (spans != null && spans.size() >= 2) {   // always contains '```aya' and '```'
+          var inner = ImmutableSeq.from(spans).view().drop(1).dropLast(1).toImmutableSeq();
+          // remove the last line break if not empty
+          if (!raw.isEmpty()) raw = raw.substring(0, raw.length() - 1);
+          yield new Literate.CodeBlock(fromSourceSpans(inner), language, raw);
+        }
+        throw new InternalException("SourceSpans");
+      }
+      case Code inlineCode -> {
+        var spans = inlineCode.getSourceSpans();
+        if (spans != null && spans.size() == 1) {
+          var sourceSpan = spans.get(0);
+          var lineIndex = linesIndex.get(sourceSpan.getLineIndex());
+          var startFrom = lineIndex + sourceSpan.getColumnIndex();
+          var sourcePos = fromSourceSpans(file, startFrom, Seq.of(sourceSpan));
+          assert sourcePos != null;
+          yield CodeOptions.analyze(inlineCode, producer.expr(inlineCode.getLiteral(), sourcePos));
+        }
+        throw new InternalException("SourceSpans");
+      }
+      default -> {
+        var spans = node.getSourceSpans();
+        if (spans == null) throw new InternalException("SourceSpans");
+        var pos = fromSourceSpans(Seq.from(spans));
+        if (pos == null) throw new UnsupportedOperationException("TODO: Which do the nodes have not source spans?");
+        producer.reporter().report(new UnsupportedMarkdown(pos, node.getClass().getSimpleName()));
+        yield new Literate.Unsupported(mapChildren(node, producer));
+      }
+    };
+  }
+
+  public @Nullable SourcePos fromSourceSpans(@NotNull Seq<SourceSpan> sourceSpans) {
+    if (sourceSpans.isEmpty()) return null;
+    var startFrom = linesIndex.get(sourceSpans.first().getLineIndex());
+
+    return fromSourceSpans(file, startFrom, sourceSpans);
+  }
+
+  /**
+   * Build a {@link SourcePos} from a list of {@link SourceSpan}
+   *
+   * @param startFrom   the SourcePos should start from. (inclusive)
+   * @param sourceSpans a not null sequence
+   * @return null if an empty sourceSpans
+   */
+  @Contract(pure = true) public static @Nullable SourcePos
+  fromSourceSpans(@NotNull SourceFile file, int startFrom, @NotNull Seq<SourceSpan> sourceSpans) {
+    if (sourceSpans.isEmpty()) return null;
+
+    var it = sourceSpans.iterator();
+    var beginSpan = it.next();
+    var endSpan = beginSpan;
+    var endLine = beginSpan.getLineIndex();
+    var endColumn = beginSpan.getLength() - 1;
+    var totalLength = beginSpan.getLength();
+
+    while (it.hasNext()) {
+      var curSpan = it.next();
+
+      // Continuous?
+      while (endSpan.getLineIndex() + 1 != curSpan.getLineIndex()) {
```

Comment:
Here is a little bit duplicated with `FillCodeBlock.Visitor`. In fact, this is the prototype of the visitor. I didn't remove this since I don't know which nodes have discontinuous source spans except `FencedCodeBlock`.

Reply:
No IDEA 😥